File: Examples/objective-c/ProximityContent/ProximityContent/Estimote/ProximityContentManager.m
//
// Please report any problems with this app template to contact@estimote.com
//

#import "ProximityContentManager.h"

#import "BeaconContentFactory.h"
#import "NearestBeaconManager.h"

@interface ProximityContentManager () <NearestBeaconManagerDelegate>

@property (nonatomic) id<BeaconContentFactory> beaconContentFactory;
@property (nonatomic) NearestBeaconManager *nearestBeaconManager;

@end

@implementation ProximityContentManager

- (instancetype)initWithBeaconRegions:(NSArray *)beaconRegions beaconContentFactory:(id<BeaconContentFactory>)beaconContentFactory {
    self = [super init];
    if (self) {
        self.beaconContentFactory = beaconContentFactory;
        self.nearestBeaconManager = [[NearestBeaconManager alloc] initWithBeaconRegions:beaconRegions];
        self.nearestBeaconManager.delegate = self;
    }
    return self;
}

- (instancetype)initWithBeaconIDs:(NSArray *)beaconIDs beaconContentFactory:(id<BeaconContentFactory>)beaconContentFactory {
    NSMutableArray *beaconRegions = [[NSMutableArray alloc] initWithCapacity:beaconIDs.count];
    for (BeaconID *beaconID in beaconIDs) {
        [beaconRegions addObject:beaconID.asBeaconRegion];
    }
    return [self initWithBeaconRegions:beaconRegions beaconContentFactory:beaconContentFactory];
}

- (void)startContentUpdates {
    [self.nearestBeaconManager startNearestBeaconUpdates];
}

- (void)stopContentUpdates {
    [self.nearestBeaconManager stopNearestBeaconUpdates];
}

- (void)nearestBeaconManager:(NearestBeaconManager *)nearestBeaconManager didUpdateNearestBeaconID:(BeaconID *)beaconID {
    if (beaconID) {
        [self.beaconContentFactory contentForBeaconID:beaconID completion:^(id content) {
            [self.delegate proximityContentManager:self didUpdateContent:content];
        }];
    } else {
        [self.delegate proximityContentManager:self didUpdateContent:nil];
    }
}

@end

File: Examples/swift/ProximityContent/ProximityContent/Estimote/ProximityContentManager.swift
//
// Please report any problems with this app template to contact@estimote.com
//

protocol ProximityContentManagerDelegate: class {

    func proximityContentManager(_ proximityContentManager: ProximityContentManager, didUpdateContent content: AnyObject?)

}

class ProximityContentManager: NearestBeaconManagerDelegate {

    weak var delegate: ProximityContentManagerDelegate?

    private let beaconContentFactory: BeaconContentFactory

    private let nearestBeaconManager: NearestBeaconManager

    init(beaconRegions: [CLBeaconRegion], beaconContentFactory: BeaconContentFactory) {
        self.beaconContentFactory = beaconContentFactory
        self.nearestBeaconManager = NearestBeaconManager(beaconRegions: beaconRegions)
        self.nearestBeaconManager.delegate = self
    }

    convenience init(beaconIDs: [BeaconID], beaconContentFactory: BeaconContentFactory) {
        let beaconRegions = beaconIDs.map { $0.asBeaconRegion }
        self.init(beaconRegions: beaconRegions, beaconContentFactory: beaconContentFactory)
    }

    func startContentUpdates() {
        self.nearestBeaconManager.startNearestBeaconUpdates()
    }

    func stopContentUpdates() {
        self.nearestBeaconManager.stopNearestBeaconUpdates()
    }

    func nearestBeaconManager(_ nearestBeaconManager: NearestBeaconManager, didUpdateNearestBeaconID nearestBeaconID: BeaconID?) {
        if let nearestBeaconID = nearestBeaconID {
            self.beaconContentFactory.requestContent(for: nearestBeaconID) { (proximityContent) in
                self.delegate?.proximityContentManager(self, didUpdateContent: proximityContent)
            }
        } else {
            self.delegate?.proximityContentManager(self, didUpdateContent: nil)
        }
    }

}
